File: tools/pe-section-analyzer.py
#!/usr/bin/env python3
from __future__ import annotations
"""Analyze PE sections for packing, encryption, and anomalies.

Usage:
    python3 pe-section-analyzer.py <binary>
    python3 pe-section-analyzer.py --entropy <binary>    # show per-section entropy
    python3 pe-section-analyzer.py --json <binary>

Checks: section entropy (high = encrypted/compressed), virtual-vs-raw size ratio
(packed sections), anomalous characteristics, empty sections, overlays.
"""
import sys
import struct
import os
import math
import json
import argparse
import subprocess


def entropy(data: bytes) -> float:
    """Shannon entropy of a byte sequence (0.0 = uniform, 8.0 = random)."""
    if not data:
        return 0.0
    freq = [0] * 256
    for b in data:
        freq[b] += 1
    length = len(data)
    ent = 0.0
    for f in freq:
        if f > 0:
            p = f / length
            ent -= p * math.log2(p)
    return ent


# ── PE parser imports ─────────────────────────────────────────────────────────
import os
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from tools.lib.pe_parse import parse_pe

def analyze(filepath):
    with open(filepath, 'rb') as f:
        data = f.read()

    try:
        pe = parse_pe(data)
    except SystemExit as e:
        return {'error': str(e)}

    sections = []
    for s in pe['sections']:
        sec_data = data[s['raddr']:s['raddr'] + s['rsize']] if s['rsize'] > 0 else b''
        ent = entropy(sec_data)

        chars = s['chars']
        flags = []
        if chars & 0x00000020:
            flags.append('CODE')
        if chars & 0x00000040:
            flags.append('IDATA')
        if chars & 0x00000080:
            flags.append('UDATA')
        if chars & 0x20000000:
            flags.append('EXEC')
        if chars & 0x40000000:
            flags.append('READ')
        if chars & 0x80000000:
            flags.append('WRITE')

        anomalies = []
        if ent > 7.0 and s['rsize'] > 1024:
            anomalies.append('HIGH ENTROPY (encrypted/compressed)')
        if ent < 1.0 and s['rsize'] > 1024:
            anomalies.append('VERY LOW ENTROPY (padding/zeros)')
        if s['rsize'] == 0 and s['vsize'] > 0:
            anomalies.append('EMPTY ON DISK (runtime-unpacked)')
        if s['vsize'] > 0 and s['rsize'] > 0 and s['vsize'] > s['rsize'] * 5:
            anomalies.append(f'PACKED (VS/RS ratio: {s["vsize"]/s["rsize"]:.1f}x)')
        if 'EXEC' in flags and 'WRITE' in flags:
            anomalies.append('WRITABLE+EXECUTABLE (SMC / packer)')
        name = s['name']
        if name.startswith('.') and name[1:].isdigit():
            anomalies.append('NUMERIC SECTION NAME (obfuscator)')

        sections.append({
            'name': name,
            'virtual_addr': s['vaddr'],
            'virtual_size': s['vsize'],
            'raw_addr': s['raddr'],
            'raw_size': s['rsize'],
            'entropy': round(ent, 3),
            'flags': flags,
            'anomalies': anomalies,
        })

    # Overlay
    overlay_size = 0
    if sections:
        last_end = max(s['raw_addr'] + s['raw_size'] for s in sections)
        if last_end < len(data):
            overlay_size = len(data) - last_end

    return {
        'file': os.path.basename(filepath),
        'size': len(data),
        'arch': 'x64' if pe['pe64'] else 'x86',
        'sections': sections,
        'overlay_size': overlay_size,
    }


def main():
    base_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    bin_name = 'pe-section-analyzer.exe' if os.name == 'nt' else 'pe-section-analyzer'
    binary_path = os.path.join(base_dir, 'bin', bin_name)
    if os.path.exists(binary_path):
        try:
            res = subprocess.run([binary_path] + sys.argv[1:])
            sys.exit(res.returncode)
        except Exception:
            pass

    parser = argparse.ArgumentParser(description='PE section analyzer')
    parser.add_argument('binary', help='PE binary to analyze')
    parser.add_argument('--entropy', action='store_true', help='Show entropy bar graph')
    parser.add_argument('--json', action='store_true', help='JSON output')
    args = parser.parse_args()

    result = analyze(args.binary)

    if args.json:
        print(json.dumps(result, indent=2))
        return

    if 'error' in result:
        print(f"Error: {result['error']}")
        sys.exit(1)

    print(f"File: {result['file']} ({result['size']:,} bytes, {result['arch']})")
    print()

    hdr = f"{'Name':<12s} {'VAddr':>10s} {'VSize':>10s} {'RSize':>10s} {'Entropy':>8s} {'Flags':<24s} Anomalies"
    print(hdr)
    print('─' * len(hdr))

    for s in result['sections']:
        flags_str = ','.join(s['flags'])
        anom_str = '; '.join(s['anomalies']) if s['anomalies'] else '—'

        ent_str = f"{s['entropy']:.3f}"
        if args.entropy:
            bar_len = int(s['entropy'] * 3)  # 0-24 chars for 0-8.0 entropy
            bar = '█' * bar_len + '░' * (24 - bar_len)
            ent_str = f"{s['entropy']:.3f} {bar}"

        print(f"{s['name']:<12s} 0x{s['virtual_addr']:08X} 0x{s['virtual_size']:08X} "
              f"0x{s['raw_size']:08X} {ent_str:>8s} {flags_str:<24s} {anom_str}")

    if result['overlay_size'] > 0:
        print(f"\nOverlay: {result['overlay_size']:,} bytes after last section")

    # Summary
    anomalous = [s for s in result['sections'] if s['anomalies']]
    if anomalous:
        print(f"\n⚠ {len(anomalous)} section(s) with anomalies detected")


if __name__ == '__main__':
    main()

File: tools/pe-parser/src/bin/pe-section-analyzer.rs
use clap::Parser;
use pe_parser::{load_binary_data, parse_pe_bytes, section_data};
use serde::Serialize;
use std::path::PathBuf;

#[derive(Parser)]
#[command(about = "Analyze binary sections for packing, entropy, and anomalies")]
struct Args {
    binary: PathBuf,
    #[arg(long)]
    entropy: bool,
    #[arg(long)]
    json: bool,
}

#[derive(Serialize)]
struct SectionReport {
    name: String,
    virtual_addr: u32,
    virtual_size: u32,
    raw_addr: u32,
    raw_size: u32,
    entropy: f64,
    flags: Vec<String>,
    anomalies: Vec<String>,
}

#[derive(Serialize)]
struct Report {
    file: String,
    size: usize,
    arch: String,
    sections: Vec<SectionReport>,
    overlay_size: usize,
}

fn entropy(data: &[u8]) -> f64 {
    if data.is_empty() {
        return 0.0;
    }
    let mut freq = [0usize; 256];
    for byte in data {
        freq[*byte as usize] += 1;
    }
    let len = data.len() as f64;
    let mut ent = 0.0;
    for count in freq {
        if count > 0 {
            let p = count as f64 / len;
            ent -= p * p.log2();
        }
    }
    (ent * 1000.0).round() / 1000.0
}

fn basename(path: &PathBuf) -> String {
    path.file_name()
        .and_then(|name| name.to_str())
        .unwrap_or_else(|| path.to_str().unwrap_or(""))
        .to_string()
}

fn analyze(path: &PathBuf) -> Result<Report, String> {
    let data = load_binary_data(path)?;
    let pe = parse_pe_bytes(&data)?;
    let mut sections = Vec::new();

    for sec in &pe.sections {
        let blob = section_data(&data, sec);
        let ent = entropy(blob);
        let mut flags = Vec::new();
        if sec.chars & 0x00000020 != 0 {
            flags.push("CODE".to_string());
        }
        if sec.chars & 0x00000040 != 0 {
            flags.push("IDATA".to_string());
        }
        if sec.chars & 0x00000080 != 0 {
            flags.push("UDATA".to_string());
        }
        if sec.chars & 0x20000000 != 0 {
            flags.push("EXEC".to_string());
        }
        if sec.chars & 0x40000000 != 0 {
            flags.push("READ".to_string());
        }
        if sec.chars & 0x80000000 != 0 {
            flags.push("WRITE".to_string());
        }

        let mut anomalies = Vec::new();
        if ent > 7.0 && sec.rsize > 1024 {
            anomalies.push("HIGH ENTROPY (encrypted/compressed)".to_string());
        }
        if ent < 1.0 && sec.rsize > 1024 {
            anomalies.push("VERY LOW ENTROPY (padding/zeros)".to_string());
        }
        if sec.rsize == 0 && sec.vsize > 0 {
            anomalies.push("EMPTY ON DISK (runtime-unpacked)".to_string());
        }
        if sec.vsize > 0 && sec.rsize > 0 && sec.vsize > sec.rsize * 5 {
            anomalies.push(format!(
                "PACKED (VS/RS ratio: {:.1}x)",
                sec.vsize as f64 / sec.rsize as f64
            ));
        }
        if flags.iter().any(|f| f == "EXEC") && flags.iter().any(|f| f == "WRITE") {
            anomalies.push("WRITABLE+EXECUTABLE (SMC / packer)".to_string());
        }
        if sec.name.starts_with('.') && sec.name[1..].chars().all(|c| c.is_ascii_digit()) {
            anomalies.push("NUMERIC SECTION NAME (obfuscator)".to_string());
        }

        sections.push(SectionReport {
            name: sec.name.clone(),
            virtual_addr: sec.vaddr,
            virtual_size: sec.vsize,
            raw_addr: sec.raddr,
            raw_size: sec.rsize,
            entropy: ent,
            flags,
            anomalies,
        });
    }

    let overlay_size = sections
        .iter()
        .map(|sec| sec.raw_addr as usize + sec.raw_size as usize)
        .max()
        .filter(|end| *end < data.len())
        .map(|end| data.len() - end)
        .unwrap_or(0);

    Ok(Report {
        file: basename(path),
        size: data.len(),
        arch: if pe.pe64 { "x64" } else { "x86" }.to_string(),
        sections,
        overlay_size,
    })
}

fn print_report(report: &Report, show_entropy_bar: bool) {
    println!(
        "File: {} ({} bytes, {})",
        report.file, report.size, report.arch
    );
    println!();
    let hdr = format!(
        "{:<12} {:>10} {:>10} {:>10} {:>8} {:<24} Anomalies",
        "Name", "VAddr", "VSize", "RSize", "Entropy", "Flags"
    );
    println!("{hdr}");
    println!("{}", "-".repeat(hdr.len()));
    for sec in &report.sections {
        let mut ent = format!("{:.3}", sec.entropy);
        if show_entropy_bar {
            let bar_len = (sec.entropy * 3.0).floor() as usize;
            ent = format!(
                "{:.3} {}{}",
                sec.entropy,
                "#".repeat(bar_len.min(24)),
                ".".repeat(24usize.saturating_sub(bar_len.min(24)))
            );
        }
        println!(
            "{:<12} 0x{:08X} 0x{:08X} 0x{:08X} {:>8} {:<24} {}",
            sec.name,
            sec.virtual_addr,
            sec.virtual_size,
            sec.raw_size,
            ent,
            sec.flags.join(","),
            if sec.anomalies.is_empty() {
                "-".to_string()
            } else {
                sec.anomalies.join("; ")
            }
        );
    }
    if report.overlay_size > 0 {
        println!(
            "\nOverlay: {} bytes after last section",
            report.overlay_size
        );
    }
    let anomalous = report
        .sections
        .iter()
        .filter(|sec| !sec.anomalies.is_empty())
        .count();
    if anomalous > 0 {
        println!("\nWARN: {anomalous} section(s) with anomalies detected");
    }
}

fn main() {
    let args = Args::parse();
    let report = match analyze(&args.binary) {
        Ok(report) => report,
        Err(e) => {
            eprintln!("Error: {e}");
            std::process::exit(1);
        }
    };
    if args.json {
        println!("{}", serde_json::to_string_pretty(&report).unwrap());
        return;
    }
    print_report(&report, args.entropy);
}
